## 评估项目数据
收集上述数据的每个内容后，从视觉上和程序上，对质量和清洁度进行数据评估。在你的 wrangle_act.ipynb中查找和记录至少 8 个质量问题 和 2 个清洁度问题。关键要点：

- 我们只需要含有图片的原始评级 (不包括转发)。
- 充分评估和清洗整个数据集需要巨大努力，所以只有一些问题 (至少 8 个质量问题和 2 个清洁度问题) 的子集需要进行评估和清洗。
- 根据清洗数据的规则，清洗包括合并数据的独立内容。
- 如果分子评级超过分母评级，不需要进行清洗。这个 特殊评级系统 是 WeRateDogs 人气度较高的主要原因。


## 清洗项目数据
评估时清洗你记录的每个问题。在 wrangle_act.ipynb 完成清洗。结果应该为优质干净的主要 pandas DataFrame (如有，或为多个 DataFrame)。必须评估符合项目动机的问题。


## 存储、分析和可视化项目数据
在 CSV 文件中存储洁净的数据，命名为 twitter_archive_master.csv。如果因为清洁需要多个表格，存在附加文件，要给这些文件合理命名。另外，你可以把清洗后的数据存储在 SQLite 数据库中 (如有需要也可以提交)。

## 在 wrangle_act.ipynb 中对清洗后的数据进行分析和可视化。必须生成至少 3 个见解和 1 个可视化。

## 项目汇报
1. 创建一个 300-600 字书面报告 命名为 wrangle_report.pdf，可以简要描述你的清洗过程。这可以作为内部文档。

2. 创建一个 250 字以上的书面报告 命名为 act_report.pdf，可以沟通观点，展示你清洗过数据后生成的可视化内容。这可作为外部文档，如博客帖子或杂志文章。

In [1]:
import numpy as np
import pandas as pd

In [3]:
## download the image-predictions.tsv
url = 'https://raw.githubusercontent.com/udacity/new-dand-advanced-china/master/%E6%95%B0%E6%8D%AE%E6%B8%85%E6%B4%97/WeRateDogs%E9%A1%B9%E7%9B%AE/image-predictions.tsv'

In [4]:
import requests
import os

file = url.split('/')[-1]
path = './WeRateDogs_data/'
r = requests.get(url, stream=True)

if os.path.isfile(path + file):
    print("File {} exist!".format(file))
    #http://docs.python-requests.org/en/master/user/quickstart/#make-a-request
else:
    with(open(path + file, 'ab')) as fd:
        for chunk in r.iter_content(chunk_size=1024):
            if chunk:
                fd.write(chunk)

File image-predictions.tsv exist!


In [2]:
!ls ./WeRateDogs_data/

image-predictions.tsv
tweet_json.txt
twitter-archive-enhanced.csv


## file 1

In [2]:
df_tweet = pd.read_json('./WeRateDogs_data/tweet_json.txt', lines=True)
df_tweet.head()

contributors  coordinates          created_at display_text_range  \
0           NaN          NaN 2017-08-01 16:23:56            [0, 85]   
1           NaN          NaN 2017-08-01 00:17:27           [0, 138]   
2           NaN          NaN 2017-07-31 00:18:03           [0, 121]   
3           NaN          NaN 2017-07-30 15:58:51            [0, 79]   
4           NaN          NaN 2017-07-29 16:00:24           [0, 138]   

                                            entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions...   
1  {'hashtags': [], 'symbols': [], 'user_mentions...   
2  {'hashtags': [], 'symbols': [], 'user_mentions...   
3  {'hashtags': [], 'symbols': [], 'user_mentions...   
4  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   

                                   extended_entities  favorite_count  \
0  {'media': [{'id': 892420639486877696, 'id_str'...           39492   
1  {'media': [{'id': 892177413194625024, 'id_str'...           33786   
2  {'media': [{'id': 891815175371796480, 'id_str'...           25445   
3  {'media': [{'id': 891689552724799489, 'id_str'...           42863   
4  {'media': [{'id': 891327551943041024, 'id_str'...           41016   

   favorited                                          full_text  geo  \
0      False  This is Phineas. He's a mystical boy. Only eve...  NaN   
1      False  This is Tilly. She's just checking pup on you....  NaN   
2      False  This is Archie. He is a rare Norwegian Pouncin...  NaN   
3      False  This is Darla. She commenced a snooze mid meal...  NaN   
4      False  This is Franklin. He would like you to stop ca...  NaN   

                         ...                          \
0                        ...                           
1                        ...                           
2                        ...                           
3                        ...                           
4                        ...                           

   possibly_sensitive_appealable  quoted_status quoted_status_id  \
0                            0.0            NaN              NaN   
1                            0.0            NaN              NaN   
2                            0.0            NaN              NaN   
3                            0.0            NaN              NaN   
4                            0.0            NaN              NaN   

   quoted_status_id_str  retweet_count  retweeted  retweeted_status  \
0                   NaN           8842      False               NaN   
1                   NaN           6480      False               NaN   
2                   NaN           4301      False               NaN   
3                   NaN           8925      False               NaN   
4                   NaN           9721      False               NaN   

                                              source truncated  \
0  <a href="http://twitter.com/download/iphone" r...     False   
1  <a href="http://twitter.com/download/iphone" r...     False   
2  <a href="http://twitter.com/download/iphone" r...     False   
3  <a href="http://twitter.com/download/iphone" r...     False   
4  <a href="http://twitter.com/download/iphone" r...     False   

                                                user  
0  {'id': 4196983835, 'id_str': '4196983835', 'na...  
1  {'id': 4196983835, 'id_str': '4196983835', 'na...  
2  {'id': 4196983835, 'id_str': '4196983835', 'na...  
3  {'id': 4196983835, 'id_str': '4196983835', 'na...  
4  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[5 rows x 31 columns]

In [4]:
df_tweet.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2352 entries, 0 to 2351
Data columns (total 31 columns):
contributors                     0 non-null float64
coordinates                      0 non-null float64
created_at                       2352 non-null datetime64[ns]
display_text_range               2352 non-null object
entities                         2352 non-null object
extended_entities                2073 non-null object
favorite_count                   2352 non-null int64
favorited                        2352 non-null bool
full_text                        2352 non-null object
geo                              0 non-null float64
id                               2352 non-null int64
id_str                           2352 non-null int64
in_reply_to_screen_name          78 non-null object
in_reply_to_status_id            78 non-null float64
in_reply_to_status_id_str        78 non-null float64
in_reply_to_user_id              78 non-null float64
in_reply_to_user_id_str          78 n

In [32]:
df_tweet['geo'].sample(3) # No use for this time

1448   NaN
485    NaN
1498   NaN
Name: geo, dtype: float64

In [5]:
df_tweet['user'].sample(3)

2094    {'id': 4196983835, 'id_str': '4196983835', 'na...
1567    {'id': 4196983835, 'id_str': '4196983835', 'na...
2296    {'id': 4196983835, 'id_str': '4196983835', 'na...
Name: user, dtype: object

In [6]:
df_tweet['extended_entities'].head()

0    {'media': [{'id': 892420639486877696, 'id_str'...
1    {'media': [{'id': 892177413194625024, 'id_str'...
2    {'media': [{'id': 891815175371796480, 'id_str'...
3    {'media': [{'id': 891689552724799489, 'id_str'...
4    {'media': [{'id': 891327551943041024, 'id_str'...
Name: extended_entities, dtype: object

In [7]:
l = df_tweet['extended_entities'].sample()

l.values.tolist()

[{'media': [{'display_url': 'pic.twitter.com/EfejX3iRGr',
    'expanded_url': 'https://twitter.com/dog_rates/status/738885046782832640/photo/1',
    'id': 738885039132401664,
    'id_str': '738885039132401664',
    'indices': [111, 134],
    'media_url': 'http://pbs.twimg.com/media/CkEMBz9WYAAGLaa.jpg',
    'media_url_https': 'https://pbs.twimg.com/media/CkEMBz9WYAAGLaa.jpg',
    'sizes': {'large': {'h': 338, 'resize': 'fit', 'w': 450},
     'medium': {'h': 338, 'resize': 'fit', 'w': 450},
     'small': {'h': 338, 'resize': 'fit', 'w': 450},
     'thumb': {'h': 150, 'resize': 'crop', 'w': 150}},
    'type': 'photo',
    'url': 'https://t.co/EfejX3iRGr'}]}]

In [8]:
l2 = df_tweet['full_text'].sample()

l2.values.tolist()

["This is Gus. He didn't win the Powerball. Quite perturbed about it. Still 10/10 would comfort in time of need https://t.co/3wc246LOtu"]

In [9]:
df_tweet['id'].head()

0    892420643555336193
1    892177421306343426
2    891815181378084864
3    891689557279858688
4    891327558926688256
Name: id, dtype: int64

In [10]:
df_tweet['retweet_count'].sample(3)

1848     767
407     4615
1810    1103
Name: retweet_count, dtype: int64

In [11]:
df_tweet['favorite_count'].sample(3)

1276     1848
687     12064
2220      262
Name: favorite_count, dtype: int64

In [12]:
df_tweet['retweet_count'].sample(3)

628      3874
1093     3099
550     18791
Name: retweet_count, dtype: int64

In [15]:
df_tweet['full_text'].sample().values.tolist()

['Meet Maggie. She can hear your cells divide. 12/10 can also probably fly https://t.co/ovE2hqXryV']

In [16]:
df_tweet['full_text'].sample().values.tolist()

["This is Kramer. He's a Picasso Tortellini. Tie couldn't be more accurate. Confident af. Runs his own business. 10/10 https://t.co/jIcVW0xxmH"]

In [3]:
var_col = ['id', 'retweet_count', 'favorite_count', 'extended_entities', 'full_text']
df_tweet_clean = df_tweet[var_col]  # do i need a copy??
df_tweet_clean.head()

id  retweet_count  favorite_count  \
0  892420643555336193           8842           39492   
1  892177421306343426           6480           33786   
2  891815181378084864           4301           25445   
3  891689557279858688           8925           42863   
4  891327558926688256           9721           41016   

                                   extended_entities  \
0  {'media': [{'id': 892420639486877696, 'id_str'...   
1  {'media': [{'id': 892177413194625024, 'id_str'...   
2  {'media': [{'id': 891815175371796480, 'id_str'...   
3  {'media': [{'id': 891689552724799489, 'id_str'...   
4  {'media': [{'id': 891327551943041024, 'id_str'...   

                                           full_text  
0  This is Phineas. He's a mystical boy. Only eve...  
1  This is Tilly. She's just checking pup on you....  
2  This is Archie. He is a rare Norwegian Pouncin...  
3  This is Darla. She commenced a snooze mid meal...  
4  This is Franklin. He would like you to stop ca...

## file 2

In [4]:
df_image = pd.read_csv('./WeRateDogs_data/image-predictions.tsv', delimiter='\t')
df_image.sample(5)

tweet_id                                          jpg_url  \
1519  787717603741622272  https://pbs.twimg.com/media/Cu6I9vvWIAAZG0a.jpg   
1360  761004547850530816  https://pbs.twimg.com/media/Co-hmcYXYAASkiG.jpg   
1214  743253157753532416  https://pbs.twimg.com/media/ClCQzFUUYAA5vAu.jpg   
883   698907974262222848  https://pbs.twimg.com/media/CbMFFssWIAAyuOd.jpg   
954   705223444686888960  https://pbs.twimg.com/media/Ccl0-HVVAAAf8aK.jpg   

      img_num                           p1   p1_conf  p1_dog  \
1519        3              German_shepherd  0.992339    True   
1360        1             golden_retriever  0.735163    True   
1214        1                     malamute  0.442612    True   
883         3  German_short-haired_pointer  0.983131    True   
954         1                 Egyptian_cat  0.090508   False   

                            p2   p2_conf  p2_dog                      p3  \
1519                  malinois  0.004920    True                  kelpie   
1360            Sussex_spaniel  0.064897    True      Labrador_retriever   
1214            Siberian_husky  0.368137    True              Eskimo_dog   
883                   bluetick  0.005558    True  curly-coated_retriever   
954   Chesapeake_Bay_retriever  0.077373    True        Mexican_hairless   

       p3_conf  p3_dog  
1519  0.000853    True  
1360  0.047704    True  
1214  0.177822    True  
883   0.003322    True  
954   0.049472    True

In [ ]:
# p1 是对推特中图片算法 #1 的预测 → basset
# p1_conf 是 #1 预测中算法的可信度 → 0.555712
# p1_dog 是 #1 预测是否是狗的品种 → True

In [19]:
df_image['img_num'].value_counts()

1    1780
2     198
3      66
4      31
Name: img_num, dtype: int64

In [5]:
df_image.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [6]:
df_image.columns

Index(['tweet_id', 'jpg_url', 'img_num', 'p1', 'p1_conf', 'p1_dog', 'p2',
       'p2_conf', 'p2_dog', 'p3', 'p3_conf', 'p3_dog'],
      dtype='object')

## file 3

In [7]:
df = pd.read_csv('./WeRateDogs_data/twitter-archive-enhanced.csv')
df.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1953  673662677122719744                    NaN                  NaN   
765   777885040357281792                    NaN                  NaN   
1933  674019345211760640                    NaN                  NaN   
1693  681297372102656000                    NaN                  NaN   
1542  689599056876867584                    NaN                  NaN   

                      timestamp  \
1953  2015-12-07 00:37:52 +0000   
765   2016-09-19 15:00:20 +0000   
1933  2015-12-08 00:15:09 +0000   
1693  2015-12-28 02:15:26 +0000   
1542  2016-01-20 00:03:21 +0000   

                                                 source  \
1953  <a href="http://twitter.com/download/iphone" r...   
765   <a href="http://twitter.com/download/iphone" r...   
1933  <a href="http://twitter.com/download/iphone" r...   
1693  <a href="http://twitter.com/download/iphone" r...   
1542  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1953  This is Kendall. 12/10 would cuddle the hell o...                  NaN   
765   This is Wesley. He's clearly trespassing. Seem...                  NaN   
1933  This is Acro. You briefly see her out of the c...                  NaN   
1693  This is actually a lion. We only rate dogs. Fo...                  NaN   
1542  Here we see 33 dogs posing for a picture. All ...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1953                       NaN                        NaN   
765                        NaN                        NaN   
1933                       NaN                        NaN   
1693                       NaN                        NaN   
1542                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1953  https://twitter.com/dog_rates/status/673662677...                12   
765   https://twitter.com/dog_rates/status/777885040...                 3   
1933  https://twitter.com/dog_rates/status/674019345...                10   
1693  https://twitter.com/dog_rates/status/681297372...                12   
1542  https://twitter.com/dog_rates/status/689599056...                11   

      rating_denominator      name doggo floofer pupper puppo  
1953                  10   Kendall  None    None   None  None  
765                   10    Wesley  None    None   None  None  
1933                  10      Acro  None    None   None  None  
1693                  10  actually  None    None   None  None  
1542                  10      None  None    None   None  None

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [9]:
df.duplicated('tweet_id').value_counts() # ID no duplicated issue.

False    2356
dtype: int64

In [32]:
id1 = int(df_tweet_clean['id'].sample().values)
print(id1)

669603084620980224


In [33]:
df_image[df_image['tweet_id'] == id1]

tweet_id                                          jpg_url  \
197  669603084620980224  https://pbs.twimg.com/media/CUroc7QW4AATIff.jpg   

     img_num           p1   p1_conf  p1_dog               p2   p2_conf  \
197        1  Maltese_dog  0.659619    True  Tibetan_terrier  0.193539   

     p2_dog        p3   p3_conf  p3_dog  
197    True  Shih-Tzu  0.039327    True

In [34]:
df[df['tweet_id'] == id1]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2155  669603084620980224                    NaN                  NaN   

                      timestamp  \
2155  2015-11-25 19:46:30 +0000   

                                                 source  \
2155  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2155  Very human-like. Cute overbite smile *finger t...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2155                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2155  https://twitter.com/dog_rates/status/669603084...                10   

      rating_denominator  name doggo floofer pupper puppo  
2155                  10  None  None    None   None  None

In [35]:
## Upper means use id can connect all three files

## Get dog rating, name and type from text

In [40]:
df[df['puppo'] == None] # all type enter is None, drop out, alse for doggo ...

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [38]:
# only keep we can use info
df_clean = df[['tweet_id', 'text',  'rating_numerator', 'rating_denominator', 'name', ]]

In [39]:
df_clean.sample(3)

tweet_id                                               text  \
787   774639387460112384  This is Sprinkles. He's trapped in light jail....   
2070  671122204919246848  Two miniature golden retrievers here. Webbed p...   
2001  672482722825261057  This is light saber pup. Ready to fight off ev...   

      rating_numerator  rating_denominator       name  
787                 10                  10  Sprinkles  
2070                 4                  10       None  
2001                10                  10      light

In [42]:
df_clean['text'].sample().values

array([ 'Have you ever seen such a smol pupper? Portable af. 12/10 would keep in shirt pocket https://t.co/KsqaIzlQ12'], dtype=object)

In [43]:
df_clean.loc[:, 'type'] = None

#dog_lists = ['pupper', 'puppo', 'doggo', 'floofer'] # typer_count =  399
dog_lists = ['pupper', 'puppo', 'doggo', 'floofer', 'blep', 'snoot'] # ref doc provide two more types

for i in range(len(df_clean)):
    text = df_clean.loc[i, 'text']
    for dog_status in dog_lists:
        if dog_status in text:
            #df_clean.type[i] = dog_status
            df_clean.loc[i, 'type'] = dog_status

/Users/lang/anaconda3/envs/ml/lib/python3.6/site-packages/pandas/core/indexing.py:337: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[key] = _infer_fill_value(value)
/Users/lang/anaconda3/envs/ml/lib/python3.6/site-packages/pandas/core/indexing.py:517: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


In [48]:
df_clean.sample(5)

tweet_id                                               text  \
591   799063482566066176  This is Mister. He only wears the most fashion...   
2081  670822709593571328  Can't do better than this lol. 10/10 for the o...   
78    877611172832227328  RT @rachel2195: @dog_rates the boyfriend and h...   
2308  666817836334096384  This is Jeph. He is a German Boston Shuttlecoc...   
62    880095782870896641  Please don't send in photos without dogs in th...   

      rating_numerator  rating_denominator    name    type  
591                 11                  10  Mister    None  
2081                10                  10    None    None  
78                  14                  10    None  pupper  
2308                 9                  10    Jeph    None  
62                  11                  10    None    None

In [49]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 6 columns):
tweet_id              2356 non-null int64
text                  2356 non-null object
rating_numerator      2356 non-null int64
rating_denominator    2356 non-null int64
name                  2356 non-null object
type                  400 non-null object
dtypes: int64(3), object(3)
memory usage: 110.5+ KB


### Clean 'name'

In [132]:
df_clean['name'].value_counts()

None              745
a                  55
Charlie            12
Oliver             11
Cooper             11
Lucy               11
Tucker             10
Penny              10
Lola               10
Bo                  9
Winston             9
Sadie               8
the                 8
Toby                7
Bailey              7
Daisy               7
Buddy               7
an                  7
Rusty               6
Oscar               6
Milo                6
Jack                6
Koda                6
Scout               6
Stanley             6
Jax                 6
Dave                6
Leo                 6
Bella               6
very                5
                 ... 
Blipson             1
Murphy              1
Major               1
Michelangelope      1
Erik                1
Vinscent            1
Sailor              1
Tessa               1
Naphaniel           1
Farfle              1
Holly               1
Stark               1
Nugget              1
Mary                1
Acro      

we can see upper list some names is incorrect:

1. None              745
2. a                  55
3. the                 8
4. an                  7
5. officially          1

In [122]:
df_clean.text.sample().values # name: Louis

array([ "This is Louis. He's crossing. It's a big deal. 13/10 h*ckin breathtaking https://t.co/D0wb1GlKAt"], dtype=object)

In [124]:
df_clean.text.sample().values # name: Shikha

array([ 'This is Shikha. She just watched you drop a skittle on the ground and still eat it. Could not be less impressed. 12/10 superior puppo https://t.co/XZlZKd73go'], dtype=object)

In [133]:
df_clean.text.sample().values # name: Brandi and Harley ?? two name

array([ 'This is Brandi and Harley. They are practicing their caroling for later. Both 12/10 festive af https://t.co/AbBDuGZUpp'], dtype=object)

In [134]:
df_clean.text.sample().values # name: swell petting?? not true name

array([ '"Thank you friend that was a swell petting" 11/10 (vid by @MatthewjamesMac) https://t.co/NY3cPAZAIM'], dtype=object)

In [79]:
name1 = df_clean.text.sample().values
print(repr(name1))

array([ 'This is a Birmingham Quagmire named Chuk. Loves to relax and watch the game while sippin on that iced mocha. 10/10 https://t.co/HvNg9JWxFt'], dtype=object)


In [69]:
import re

In [100]:
s = "hi my name is ryan, and i am new to python and would like to learn more"

# ref to: https://stackoverflow.com/questions/6633678/finding-words-after-keyword-in-python
m = re.search('(?<=name is)(.*)', s).groups()
m[0].split()[0]

'ryan,'

In [98]:
name = re.search('(?<=is)(.*)', repr(name1)).groups()
name

(" is a Birmingham Quagmire named Chuk. Loves to relax and watch the game while sippin on that iced mocha. 10/10 https://t.co/HvNg9JWxFt'], dtype=object)",)

In [ ]:
ref to :
    
    1. https://www.experts-exchange.com/questions/28079280/Finding-words-after-key-word-in-a-string-python.html

###   Assess df

In [64]:
# remove rating_numerator < 10
df2 = df[df.rating_numerator < 10]
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 440 entries, 45 to 2355
Data columns (total 17 columns):
tweet_id                      440 non-null int64
in_reply_to_status_id         12 non-null float64
in_reply_to_user_id           12 non-null float64
timestamp                     440 non-null object
source                        440 non-null object
text                          440 non-null object
retweeted_status_id           9 non-null float64
retweeted_status_user_id      9 non-null float64
retweeted_status_timestamp    9 non-null object
expanded_urls                 430 non-null object
rating_numerator              440 non-null int64
rating_denominator            440 non-null int64
name                          440 non-null object
doggo                         440 non-null object
floofer                       440 non-null object
pupper                        440 non-null object
puppo                         440 non-null object
dtypes: float64(4), int64(3), object(10)
memory us

In [65]:
df3 = df[df.rating_numerator < df.rating_denominator] 
df3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 442 entries, 45 to 2355
Data columns (total 17 columns):
tweet_id                      442 non-null int64
in_reply_to_status_id         13 non-null float64
in_reply_to_user_id           13 non-null float64
timestamp                     442 non-null object
source                        442 non-null object
text                          442 non-null object
retweeted_status_id           9 non-null float64
retweeted_status_user_id      9 non-null float64
retweeted_status_timestamp    9 non-null object
expanded_urls                 431 non-null object
rating_numerator              442 non-null int64
rating_denominator            442 non-null int64
name                          442 non-null object
doggo                         442 non-null object
floofer                       442 non-null object
pupper                        442 non-null object
puppo                         442 non-null object
dtypes: float64(4), int64(3), object(10)
memory us

In [66]:
df4 = df3[df3.rating_denominator != 10] 
df4

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
342   832088576586297345           8.320875e+17         3.058208e+07   
784   775096608509886464                    NaN                  NaN   
1068  740373189193256964                    NaN                  NaN   
1165  722974582966214656                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1662  682962037429899265                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   

                      timestamp  \
342   2017-02-16 04:45:50 +0000   
784   2016-09-11 22:20:06 +0000   
1068  2016-06-08 02:41:38 +0000   
1165  2016-04-21 02:25:47 +0000   
1274  2016-03-14 02:04:08 +0000   
1598  2016-01-10 04:04:10 +0000   
1662  2016-01-01 16:30:13 +0000   
2335  2015-11-16 16:11:11 +0000   

                                                 source  \
342   <a href="http://twitter.com/download/iphone" r...   
784   <a href="http://twitter.com/download/iphone" r...   
1068  <a href="http://twitter.com/download/iphone" r...   
1165  <a href="http://twitter.com/download/iphone" r...   
1274  <a href="http://twitter.com/download/iphone" r...   
1598  <a href="http://twitter.com/download/iphone" r...   
1662  <a href="http://twitter.com/download/iphone" r...   
2335  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
342            @docmisterio account started on 11/15/15                  NaN   
784   RT @dog_rates: After so many requests, this is...         7.403732e+17   
1068  After so many requests, this is Bretagne. She ...                  NaN   
1165  Happy 4/20 from the squad! 13/10 for all https...                  NaN   
1274  From left to right:\r\nCletus, Jerome, Alejand...                  NaN   
1598  Yes I do realize a rating of 4/20 would've bee...                  NaN   
1662  This is Darrel. He just robbed a 7/11 and is i...                  NaN   
2335  This is an Albanian 3 1/2 legged  Episcopalian...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
342                        NaN                        NaN   
784               4.196984e+09  2016-06-08 02:41:38 +0000   
1068                       NaN                        NaN   
1165                       NaN                        NaN   
1274                       NaN                        NaN   
1598                       NaN                        NaN   
1662                       NaN                        NaN   
2335                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
342                                                 NaN                11   
784   https://twitter.com/dog_rates/status/740373189...                 9   
1068  https://twitter.com/dog_rates/status/740373189...                 9   
1165  https://twitter.com/dog_rates/status/722974582...                 4   
1274  https://twitter.com/dog_rates/status/709198395...                45   
1598                                                NaN                 4   
1662  https://twitter.com/dog_rates/status/682962037...                 7   
2335  https://twitter.com/dog_rates/status/666287406...                 1   

      rating_denominator    name doggo floofer pupper puppo  
342                   15    None  None    None   None  None  
784                   11    None  None    None   None  None  
1068                  11    None  None    None   None  None  
1165                  20    None  None    None   None  None  
1274                  50    None  None    None   None  None  
1598                  20    None  None    None   None  None  
1662                  11  Darrel  None    None   None  None  
2335                   2      an  None    None   None  Non

In [18]:
df4.rating_denominator.value_counts()

11    3
20    2
15    1
2     1
50    1
Name: rating_denominator, dtype: int64

In [19]:
df3.rating_numerator.value_counts()

9     158
8     102
7      55
5      37
6      32
3      19
4      17
2       9
1       9
0       2
45      1
11      1
Name: rating_numerator, dtype: int64

In [26]:
df3[df3['rating_numerator'] == 11].text

342    @docmisterio account started on 11/15/15
Name: text, dtype: object

In [64]:
s = df3[df3['rating_numerator'] == 45]['text'].to_string()
s

'1274    From left to right:\\r\\nCletus, Jerome, Alejand...'

## 评估数据是数据整理的第二步, 
- 数据质量问题（即内容问题）
- 缺乏整洁度（即结构性问题）

这个项目就是从csv提取文本，再从文本提取评级，狗名及狗地位，
再从txt_json里至少提取转发和喜欢，
最后把提取的和image合并（inner）后评估清洗，当然先清洗后合并